pkg/apis/serving/k8s_validation_test.go

```diff
@@ -417,6 +440,122 @@ func TestPodSpecValidation(t *testing.T) {
+			Containers: []corev1.Container{{
+				Image: "busybox",
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      "foo",
+					MountPath: "/data",
+					ReadOnly:  true,
+				}},
+			}},
+			Volumes: []corev1.Volume{{
+				Name: "foo",
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: "myclaim",
+						ReadOnly:  false,
+					},
+				}},
+			}},
+		cfgOpts: []configOption{withPodSpecPersistentVolumesClaimsEnabled(), withPodSpecPersistentVolumesWriteEnabled()},
+	}, {
+		name: "PVC not read-only, write allowed, annotation missing",
+		ps: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Image: "busybox",
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      "foo",
+					MountPath: "/data",
+					ReadOnly:  true,
+				}},
+			}},
+			Volumes: []corev1.Volume{{
+				Name: "foo",
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: "myclaim",
+						ReadOnly:  false,
+					},
+				}},
+			}},
+		cfgOpts: []configOption{withPodSpecPersistentVolumesClaimsEnabled(), withPodSpecPersistentVolumesWriteAllowed()},
+		want: &apis.FieldError{
+			Message: `Persistent volumes write support is off, but found Persistent Volume claim myclaim that is not read-only`,
+		},
+	}, {
+		name: "PVC not read-only, write allowed, annotation set to enabled",
+		ps: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Image: "busybox",
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      "foo",
+					MountPath: "/data",
+					ReadOnly:  false,
+				}},
+			}},
+			Volumes: []corev1.Volume{{
+				Name: "foo",
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: "myclaim",
+						ReadOnly:  false,
+					},
+				}},
+			}},
+		cfgOpts: []configOption{withPodSpecPersistentVolumesClaimsEnabled(), withPodSpecPersistentVolumesWriteAllowed()},
+		extraCtxFunc: func(ctx context.Context) context.Context {
+			return helpers.WithAnnotations(ctx, map[string]string{PodSpecPersistenceVolumesWriteAnnotation: "enabled"})
+		},
+	}, {
+		name: "PVC not read-only, write allowed, annotation set to disabled",
+		ps: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Image: "busybox",
+				VolumeMounts: []corev1.VolumeMount{{
+					Name:      "foo",
+					MountPath: "/data",
+					ReadOnly:  true,
+				}},
+			}},
+			Volumes: []corev1.Volume{{
+				Name: "foo",
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: "myclaim",
+						ReadOnly:  false,
+					},
+				}},
+			}},
+		cfgOpts: []configOption{withPodSpecPersistentVolumesClaimsEnabled(), withPodSpecPersistentVolumesWriteAllowed()},
+		want: &apis.FieldError{
+			Message: `Persistent volumes write support is off, but found Persistent Volume claim myclaim that is not read-only`,
```

Comment:
fwiw I'm not sure I understand the point of having an annotation for this. IIUC the only case where a user should be able to set it to true is if the feature is "allowed", but if the feature is allowed why don't we just directly let them set the ReadOnly field to false without an annotation? Why would someone add ReadOnly: false and add the annotation as disabled (or, equivalently, not add it)?

Reply:
My rationale is that from an admin perspective when deploying many apps you might want to have explicit control on what apps have write access and have an easy way to mark/list them so and restrict them per case. This is due to the implications that volume write access has for the serverless model. Of course you can always implement this externally to Knative via custom tooling, annotations etc so I agree using `ReadOnly` could be enough in that case. I can remove it wdyt?

Reply:
I removed the annotation if we find it useful can be added later on.